### PR TITLE
Feature/2810 automate journals public list

### PR DESCRIPTION
**Pull request by philipkcl**
# Automate the public change log of journals added #2810

*Please don't delete any sections when completing this PR template; instead enter **N/A** for checkboxes or sections which are not applicable, unless otherwise stated below*

See #
* https://github.com/DOAJ/doajPM/issues/2810

Describe the scope/purpose of the PR here in as much detail as you like

## Categorisation

This PR...
- [ ] has scripts to run
- [x] has migrations to run
- [ ] adds new infrastructure
- [ ] changes the CI pipeline
- [ ] affects the public site
- [ ] affects the editorial area
- [ ] affects the publisher area
- [ ] affects the monitoring

## Basic PR Checklist

Instructions for developers:
* For each checklist item, if it is N/A to your PR check the N/A box
* For each item that you have done and confirmed for yourself, check Developer box (including if you have checked the N/A box)

Instructions for reviewers:
* For each checklist item that has been confirmed by the Developer, check the Reviewer box if you agree
* For multiple reviewers, feel free to add your own checkbox with your github username next to it if that helps with review tracking

### Code Style

- No deprecated methods are used
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer

- No magic strings/numbers - all strings are in `constants` or `messages` files
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer
  
- ES queries are wrapped in a Query object rather than inlined in the code
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer
  
- Where possible our common library functions have been used (e.g. dates manipulated via `dates`)
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer
  
- Cleaned up commented out code, etc
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer

- Urls are constructed with `url_for` not hard-coded
  - [x] N/A
  - [x] Developer
  - [x] Reviewer
### Testing

- Unit tests have been added/modified
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer
  
- Functional tests have been added/modified
  - [x] N/A
  - [ ] Developer
  - [x] Reviewer
  
- Code has been run manually in development, and functional tests followed locally
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer

- Have CSS/style changes been implemented?  If they are of a global scope (e.g. on base HTML elements) have the downstream impacts of the change in other areas of the system been considered?
  - [x] N/A
  - [ ] Developer
  - [x] Reviewer

### Documentation

- FeatureMap annotations have been added
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer
  
- Documentation updates - if needed - have been identified and prepared for inclusion into main documentation (e.g. added and highlighted/commented as appropriate to this PR)
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer
  
- Core model documentation has been added to if needed: https://docs.google.com/spreadsheets/d/1lun2S9vwGbyfy3WjIjgXBm05D-3wWDZ4bp8xiIYfImM/edit
  - [x] N/A
  - [ ] Developer
  - [x] Reviewer

- Events and consumers documentation has been added if needed: https://docs.google.com/spreadsheets/d/1oIeG5vg-blm2MZCE-7YhwulUlSz6TOUeY8jAftdP9JE/edit
  - [x] N/A
  - [ ] Developer
  - [x] Reviewer
  
- The docs for this branch have been generated and pushed to the doc site (see docs/README.md for details)
  - [x] N/A
  - [ ] Developer
  - [x] Reviewer

### Release Readiness

- If needed, migration has been created and tested locally
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer

- Release sheet has been created, and completed as far as is possible https://docs.google.com/spreadsheets/d/1Bqx23J1MwXzjrmAygbqlU3YHxN1Wf7zkkRv14eTVLZQ/edit
  https://docs.google.com/spreadsheets/d/1Bqx23J1MwXzjrmAygbqlU3YHxN1Wf7zkkRv14eTVLZQ/edit#gid=761616594
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer

- There has been a recent merge up from `develop` (or other base branch).  List the dates of the merges up from develop below
  - 2023-09-29

## Testing

List the Functional Tests that must be run to confirm this feature

1. ...
2. ...

## Deployment

What deployment considerations are there? (delete any sections you don't need)

### Configuration changes

What configuration changes are included in this PR, and do we need to set specific values for production
added new env variable 
```
GOOGLE_KEY_PATH = ''
DATALOG_JA_FILENAME = 'DOAJ: journals added and withdrawn'
DATALOG_JA_WORKSHEET_NAME = 'Added'
```

in order to learn how to create google api key file please reference: `docs/dev/how-to-setup.md`

```
Verify if the background job is configured in the config file.
"datalog_journal_added_update": {"month": "*", "day": "*", "day_of_week": "*", "hour": "0", "minute": "50"},
```

### Scripts

What scripts need to be run from the PR (e.g. if this is a report generating feature), and when (once, regularly, etc).

### Migrations

What migrations need to be run to deploy this
```
# Load datalog_journal_added 

load data of datalog_journal_added

### Run
python -m portality.migrate.2810_auto_publiclist_journal_added.main_upgrade publiclsit_journal_added.xlsx

publiclsit_journal_added.xlsx is the file that is exported from google sheets

```

### Monitoring

What additional monitoring is required of the application as a result of this feature

### New Infrastructure

What new infrastructure does this PR require (e.g. new services that need to run on the back-end).

### Continuous Integration

What CI changes are required for this

### CODE REVIEW

**Comment by philipkcl** [comment]
> Thanks @philipkcl Please note that we will start [a new tab](https://docs.google.com/spreadsheets/d/183mRBRqs2jOyP0qZWXN8dUd02D4vL0Mov_kgYF8HORM/edit#gid=1714479563&range=A1) in the same Google sheet and not continue with the current one since it has become too big.

@dommitchell , file name and tab name is configurable, all we need to do is update doaj config file and restart the doaj server if tab name need to be change in the future. When create a new tab, we need to make sure the new tab have the same table format (same columns) as original one.

File: portality/tasks/datalog_journal_added_update.py [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+        row = next(records, ['x'])
+        latest_row_index += 1
+        if any(r.strip() for r in row):
+            return latest_row_index
+
+
+def find_first_issn(rows):
+    for row in rows:
+        for c in row:
+            if re.match(r'[0-9X]{4}-[0-9X]{4}', c):
```
Comment (Steven-Eardley) [inline portality/tasks/datalog_journal_added_update.py:123]: I think we already have an ISSN regex in `regex.py` - can it be reused?

File: portality/settings.py [hunk 2]
```diff
@@ -439,7 +439,8 @@
     "anon_export": {"month": "*", "day": "10", "day_of_week": "*", "hour": "6", "minute": "30"},
     "old_data_cleanup": {"month": "*", "day": "12", "day_of_week": "*", "hour": "6", "minute": "30"},
     "monitor_bgjobs": {"month": "*", "day": "*/6", "day_of_week": "*", "hour": "10", "minute": "0"},
-    "find_discontinued_soon": {"month": "*", "day": "*", "day_of_week": "*", "hour": "0", "minute": "3"}
+    "find_discontinued_soon": {"month": "*", "day": "*", "day_of_week": "*", "hour": "0", "minute": "3"},
+    "datalog_journal_added_update": {"month": "*", "day": "*", "day_of_week": "*", "hour": "0", "minute": "50"},
```
Comment (Steven-Eardley) [inline portality/settings.py:443]: `test.cfg` (and our own dev.cfg files) will need a `CRON_NEVER` for this

File: setup.py [hunk 3]
```diff
@@ -55,6 +55,13 @@
         "esprit @ git+https://github.com/CottageLabs/esprit.git@edda12177effa0945d99302f0d453b22503e335b#egg=esprit",
         "dictdiffer @ git+https://github.com/CottageLabs/dictdiffer.git@cc86c1ca1a452169b1b2e4a0cb5fc9e6125bc572#egg=dictdiffer",
         "flask-swagger @ git+https://github.com/DOAJ/flask-swagger.git@f1dbf918d9903a588eed3cce2a87eeccc9f8cc0e#egg=flask-swagger"
+
+        # pri list and automate public list
+        'gspread~=5.10.0',
+        'oauth2client~=4.1.3',
+        'pandas~=2.0.1',  # pandas lets us generate URLs for linkcheck
```
Comment (Steven-Eardley) [inline setup.py:62]: I'm not seeing pandas imported anywhere - does it get installed under `gspread`?

**Review by Steven-Eardley — COMMENTED** [review]
A couple of comments about code reuse and config

File: setup.py (hunk 3 above)
Comment (philipkcl) [reply setup.py:62]: yes, it required by `gspread-dataframe`, in `extras_require.test` also using pandas, so I move it to `install_requires`  for explicit required.

**Review by Steven-Eardley — APPROVED** [review]

**Review by RK206 — APPROVED** [review]

**Pull request merged**